superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx

```diff
@@ -0,0 +1,518 @@
+import { Button } from 'react-bootstrap';
+import { t } from '@superset-ui/translation';
+import { safeStringify } from '../../../utils/safeStringify';
+
+import getFilterScopeNodesTree from '../../util/getFilterScopeNodesTree';
+import getFilterFieldNodesTree from '../../util/getFilterFieldNodesTree';
+import getFilterScopeParentNodes from '../../util/getFilterScopeParentNodes';
+import getCurrentScopeChartIds from '../../util/getCurrentScopeChartIds';
+import getRevertedFilterScope from '../../util/getRevertedFilterScope';
+import FilterScopeTree from './FilterScopeTree';
+import FilterFieldTree from './FilterFieldTree';
+import {
+  getChartIdAndColumnFromFilterKey,
+  getDashboardFilterKey,
+} from '../../util/getDashboardFilterKey';
+
+const propTypes = {
+  dashboardFilters: PropTypes.object.isRequired,
+  layout: PropTypes.object.isRequired,
+  filterImmuneSlices: PropTypes.arrayOf(PropTypes.number).isRequired,
+  filterImmuneSliceFields: PropTypes.object.isRequired,
+  setDirectPathToChild: PropTypes.func.isRequired,
+  onCloseModal: PropTypes.func.isRequired,
+};
+
+export default class FilterScopeSelector extends React.PureComponent {
+  constructor(props) {
+    super(props);
+
+    const {
+      dashboardFilters,
+      filterImmuneSlices,
+      filterImmuneSliceFields,
+      layout,
+    } = props;
+
+    if (Object.keys(dashboardFilters).length > 0) {
+      // display filter fields in tree structure
+      const filterFieldNodes = getFilterFieldNodesTree({
+        dashboardFilters,
+        isSingleEditMode: true,
+      });
+      this.allfilterFields = [];
+      filterFieldNodes.forEach(({ children }) => {
+        children.forEach(child => {
+          this.allfilterFields.push(child.value);
+        });
+      });
+
+      if (filterFieldNodes.length && filterFieldNodes[0].children) {
+        this.defaultFilterKey = filterFieldNodes[0].children[0].value;
+      }
+      const checkedFilterFields = [this.defaultFilterKey];
+      // expand defaultFilterKey
+      const { chartId } = getChartIdAndColumnFromFilterKey(
+        this.defaultFilterKey,
+      );
+      const expandedFilterIds = [chartId];
+
+      // display checkbox tree of whole dashboard layout
+      const filterScopeMap = Object.values(dashboardFilters).reduce(
+        (map, { chartId: filterId, columns }) => {
+          const filterScopeByChartId = Object.keys(columns).reduce(
+            (mapByChartId, columnName) => {
+              const filterKey = getDashboardFilterKey({
+                chartId: filterId,
+                column: columnName,
+              });
+              const nodes = getFilterScopeNodesTree({
+                components: layout,
+                isSingleEditMode: true,
+                checkedFilterFields,
+                selectedChartId: filterId,
+              });
+              const checked = getCurrentScopeChartIds({
+                scopeComponentIds: ['ROOT_ID'], // dashboardFilters[chartId].scopes[columnName],
```

Comment:
should this always be `'ROOT_ID'`? or is the commented part here supposed to be uncommented?

Reply:
by default, or all new filter field, their scope should be all dashboard charts. But after user change/set new scope for a filter field, the new scope object will be stored in redux state (and saved to dashboard metadata in backend).
When user open filter scope selector, the saved filter scope will be pre-populated from redux store. This line is mock data for this component. It will be changed in the next PR.